packages/core/src/shared/crashMonitoring.ts

```diff
@@ -324,6 +359,84 @@ class CrashChecker {
 
     /** Mimics a crash, only for testing */
     public testCrash() {
+        this.cleanup()
+    }
+}
+
+/**
+ * This class is required for the following edge case:
+ * 1. Heartbeat is sent
+ * 2. Computer goes to sleep for X minutes
+ * 3. Wake up computer. But before a new heartbeat can be sent, a crash checker (can be from another ext instance) runs
+ *   and sees a stale heartbeat. It assumes a crash.
+ *
+ * Why? Intervals do not run while the computer is asleep, so the latest heartbeat has a "lag" since it wasn't able to send
+ *      a new heartbeat.
+ *      Then on wake, there is a racecondition for the next heartbeat to be sent before the next crash check. If the crash checker
+ *      runs first it will incorrectly conclude a crash.
+ *
+ * Solution: Keep track of the lag, and then skip the next crash check if there was a lag. This will give time for the
+ *           next heartbeat to be sent.
+ */
+class TimeLag {
```

Comment:
what is this class abstracting? why is this not part of the existing Heartbeat class?

Reply:
The underlying issue is **checking** for a stale heartbeat. We cannot reliably compensate for lag by sending a heartbeat since the crash checking is a separate process, and there becomes a race condition between sending a fresh heartbeat vs checking for a stale one. The solution is that the crash checker compensates for "time lag" by skipping a check

This class could live inside the CrashChecker class, but I'm currently passing the TimeLag instance in as a dependency to it.